### PR TITLE
Refactor the attention metadata into separate classes

### PR DESCRIPTION
**Pull request by kanz-nv**
# What does this PR do ?
Refactors the attention metadata into separate classes.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Review by shanmugamr1992 — APPROVED** [review]

**Comment by kvareddy** [comment]
@lmcafee-nvidia can you please take a look at this MR.

File: megatron/core/inference/contexts/attention_context/metadata_base.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+import logging
+
+
+class MetadataBase:
+    """
+    Base class for attention metadata.
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/attention_context/metadata_base.py:6]: Can you add a bit more info on what this metadata is being used for, and generally what kind of data should be stored in metadata? For anyone looking at this file for the first time, they won't know what this is for.

File: megatron/core/inference/contexts/attention_context/metadata_base.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Reset the metadata.
+        """
+        pass
+
+    def tensor_copy_and_pad(
+        self,
+        tensor_buf,
+        tensor,
+        real_batch_size,
+        graph_batch_size,
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/attention_context/metadata_base.py]: general comments on naming clarity here:
- the name `graph_batch_size` is a bit confusing in a method called `tensor_copy_and_pad` (even though someone knowledgeable might know this is used for graph padding), since graphs aren't otherwise involved in this method. You might consider renaming the batch sizes to `unpadded_batch_size` and `padded_batch_size`.
- I think the namings should be aligned between tensor vars and batch size vars. E.g., `unpadded_tensor` & `unpadded_batch_size`, and `padded_tensor` & `padded_batch_size`

File: megatron/core/inference/contexts/attention_context/metadata_base.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        Copy the unpadded tensor to the tensor_buf,
+        pad the tensor_buf with zero or the last value of the tensor,
+        depending on whether the tensor is cumulative.
+        Args:
+            tensor_buf: The destination tensor, at least padded_batch_size long.
+            unpadded_tensor: The tensor to copy, at least real_batch_size long.
+            real_batch_size: The real batch size.
+            padded_batch_size: Padded boundary of the tensor.
+            is_cumulative_tensor: Whether the tensor is cumulative.
+                If True, we pad the tensor_buf with the last value of the unpadded_tensor.
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/contexts/attention_context/metadata_base.py:51]: missing argument description for pad_value.

File: megatron/core/inference/contexts/attention_context/mha_metadata.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        self._cu_kv_seq_lengths_buf = torch.zeros(self.max_bs + 1, dtype=torch.int32, device=device)
+        self._kv_seq_lengths_buf = torch.zeros(self.max_bs, dtype=torch.int32, device=device)
+        self._block_table_buf = torch.zeros(
+            (self.max_bs, self.max_kv_blocks), dtype=torch.int32, device=device
+        )
+        self._max_seqlen_q = 0
+        self._max_seqlen_k = 0
+        self.state_data = {}
+
+    def update(
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/contexts/attention_context/mha_metadata.py:38]: is this `update` method analogous to the old `initialize_attention_states`?

File: megatron/core/inference/contexts/attention_context/mha_metadata.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        self._query_lengths_buf.fill_(0)
+        self._cu_query_seq_lengths_buf.fill_(0)
+        self._cu_kv_seq_lengths_buf.fill_(0)
+        self._kv_seq_lengths_buf.fill_(0)
+        self._block_table_buf.fill_(0)
+        self._max_seqlen_q = 0
+        self._max_seqlen_k = 0
+
+
+class GraphMHAMetadata(MHAMetadata):
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/contexts/attention_context/mha_metadata.py]: nit: `CudaGraph` or `CudaGraphed` instead of just Graph?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 6]
```diff
@@ -881,89 +884,29 @@ def initialize_attention_state(
...
-            ] = request_to_kv_block_ids
-            self.block_table = self.request_to_kv_block_ids_cudagraph_only[
-                : self.padded_active_request_count
-            ]
-        else:
-            self.block_table = self.request_to_kv_block_ids[
-                self.paused_request_count : self.total_request_count
-            ]
+        # Update cu_query_seq_lengths, max_seqlen_q.
+        active_slice = slice(self.paused_request_count, self.total_request_count)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:943]: I did not know there's a python function called slice to represent sliced indices, nice!

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]
LGTM overall -  My only ask would be to use a better name than `update` - In my mental model, we have been using `update_requests` to represent everything that happens after an engine step. 
Maybe just reuse the old `initialize_attention_states` name? or maybe just `initiaize` or `initialize_states`

**Review by shanmugamr1992 — APPROVED** [review]

**Review by sidsingh-nvidia — APPROVED** [review]
